### ISSUE TITLE
node-test-commit-osx is perma-red with git failures

### ISSUE THREAD

**Issue by Trott** [nodejs/build#839]
node-test-commit-osx is perma-red with git failures. @nodejs/build 

Host is test-requireio_andineck-debian7-arm_pi1p-1 which seems odd because...os x.  See https://ci.nodejs.org/job/node-test-commit-osx/11862/ for an example. Or  here's some sample output:

```console
Started by upstream project "node-test-commit" build number 11906
originally caused by:
 Started by upstream project "node-test-pull-request" build number 9750
 originally caused by:
  Started by user Rich Trott
[EnvInject] - Loading node environment variables.
Building remotely on test-requireio_andineck-debian7-arm_pi1p-1 (pi1-raspbian-wheezy) in workspace /home/iojs/build/workspace/node-test-commit-osx@2
 > git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > git config remote.origin.url https://github.com/nodejs/node.git # timeout=10
Fetching upstream changes from https://github.com/nodejs/node.git
 > git --version # timeout=10
using GIT_SSH to set credentials 
 > git fetch --tags --progress https://github.com/nodejs/node.git +refs/heads/*:refs/remotes/origin/* +refs/pull/14926/head:refs/remotes/origin/_jenkins_local_branch # timeout=20
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from https://github.com/nodejs/node.git
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:817)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1084)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1115)
	at hudson.scm.SCM.checkout(SCM.java:496)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1281)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:604)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:529)
	at hudson.model.Run.execute(Run.java:1728)
	at hudson.matrix.MatrixBuild.run(MatrixBuild.java:313)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:405)
Caused by: hudson.plugins.git.GitException: Command "git fetch --tags --progress https://github.com/nodejs/node.git +refs/heads/*:refs/remotes/origin/* +refs/pull/14926/head:refs/remotes/origin/_jenkins_local_branch" returned status code 128:
stdout: 
stderr: fatal: The remote end hung up unexpectedly

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1924)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1643)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:71)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:352)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:153)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$1.call(RemoteGitImpl.java:146)
	at hudson.remoting.UserRequest.perform(UserRequest.java:153)
	at hudson.remoting.UserRequest.perform(UserRequest.java:50)
	at hudson.remoting.Request$2.run(Request.java:336)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:68)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at hudson.remoting.Engine$1$1.run(Engine.java:94)
	at java.lang.Thread.run(Thread.java:744)
	at ......remote call to JNLP4-connect connection from static-n49-176-213-208.sbr2.nsw.optusnet.com.au/49.176.213.208:34186(Native Method)
	at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1545)
	at hudson.remoting.UserResponse.retrieve(UserRequest.java:253)
	at hudson.remoting.Channel.call(Channel.java:830)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:146)
	at sun.reflect.GeneratedMethodAccessor675.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:132)
	at com.sun.proxy.$Proxy88.execute(Unknown Source)
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:815)
	... 11 more
ERROR: Error fetching remote repo 'origin'
Notifying upstream projects of job completion
Finished: FAILURE
```

**Comment by Trott** [comment]
(11 consecutive failures and counting at time of this writing, just in case that's useful.)

**Comment by refack** [comment]
I think there is something wrong with the machine config:
![image](https://user-images.githubusercontent.com/96947/29489009-96d66814-84e4-11e7-9ac4-97cc054eb904.png)
It seems like the job is trying to run on a Pi

**Comment by Trott** [comment]
Same thing going on with the fips job. @refack reminded me that @rvagg had a power outage ("Pi-tastrophe" as @refack said) and it seems maybe an errant click or something in the cleanup after that maybe caused this or something like that?

**Comment by gibfahn** [comment]
>It seems like the job is trying to run on a Pi

So the job running on a PI is fine, that's the parent job, which just clones the repo and then triggers the actual build.

This looks like an issue with that machine.

**Comment by joaocgreis** [comment]
As @gibfahn mentioned, the `test-osx` job, like many others, is a multi-configuration job: there is something like a main/parent job that spawns sub-jobs that run on the actual platforms that we want to test (when there is only one, like in the OSX job, Jenkins calls it `default` instead of the platform label). The problem was that Jenkins was changed a few days ago to run all of the parent jobs in dedicated machines instead of the Jenkins master. Jenkins offers two mechanisms to bind jobs to a specific machine: "Preference of node" and "Restrict where this project can be run" (under advanced). We have always been using "Preference of node" without issue. However, now, probably because it's no longer `master`, Jenkins is not respecting the preference as absolute, or for some reason it's not working as before.

I'll change all jobs to "Restrict where this project can be run". This apparently solved the issue for OSX and FIPS, the OSX machine itself seems fine.

**Comment by refack** [comment]
BTW: perfect opportunity to use the misunderstood "Resume Build" button
Original job - failed on osx and FIPS - https://ci.nodejs.org/job/node-test-commit/11909/
Rebuild - only tries those two: - https://ci.nodejs.org/job/node-test-commit/11913/
![image](https://user-images.githubusercontent.com/96947/29490973-36bf43c4-851b-11e7-9c84-137dc6fe47cb.png)

**Comment by rvagg** [comment]
I'm certain I restricted this job and fips and most of the rest of the unrestricted jobs to run on jenkins-workspace when I made the change. Have we had some kind of config reversion perhaps? I even grepped through config files to find out which ones might be impacted to do this.

**Comment by joaocgreis** [comment]
@rvagg that was using the "Preference of node" option, I confirm it was set to `jenkins-workspace`. You did change all of them. That was how it's done before for master in many jobs, and has always worked until now. (Or are you sure you changed them to "Restrict where this project can be run" under "Advanced options"?)

I'm changing all the jobs to "Restrict where this project can be run", 41 left now.

**Comment by joaocgreis** [comment]
Done. I changed 74 jobs to "Restrict where this project can be run", another 68 were already working this way. I've confirmed that `canRun` is false in all jobs that have it (in `config.xml`) - it is true when "Preference of node" is used.

@rvagg this is what I've changed: https://ci.nodejs.org/view/All/job/node-test-commit-linux-fips/jobConfigHistory/showDiffFiles?timestamp1=2017-08-19_19-45-23&timestamp2=2017-08-19_19-46-40 , you might have check some of the jobs that already had this. My guess is that you did change everything and changed everything right, but missed that there are two different options that accomplish the same (except that one only works some times).

As a last note, I found some more jobs that were running in other nodes not `jenkins-workspace`: https://ci.nodejs.org/view/All/job/node-test-commit-linux/11946/ , https://ci.nodejs.org/view/All/job/node-compile-windows/11254/ . My guess is there are plenty more, and we were bound to see this happen when they some ran in the Pis. Apparently, when jenkins decides to use some node to run a job, it'll keep using it for all runs: that's why OSX was failing every time.

**Comment by rvagg** [comment]
Excellent work @joaocgreis, thanks for doing this!

**Comment by rvagg** [comment]
@joaocgreis the coverity nightly builds were failing cause they were moved to jenkins-workspace but they depend on the coverity tools in /var/cov-analysis-linux64-2017.07/ which is installed on test-digitalocean-debian8-x64-1 which is where it was running before. I've installed these tools onto the jenkins-workspace servers so it can run there now. Unfortunately there's a compile involved and these machines aren't ideal for compiling (Atom), but they are not terrible and this isn't a time-sensitive build so I think this is OK.

The coverity tools aren't in Ansible (afaik) but I guess we should document this somewhere.
